fake/fake_clock_121.go

```diff
@@ -0,0 +1,35 @@
+//go:build go1.20
+
+package fake
+
+import (
+	"context"
+	"time"
+)
+
+func (f *Clock) ContextWithDeadlineCause(ctx context.Context, t time.Time, cause error) (context.Context, context.CancelFunc) {
+	cctx, cancelCause := context.WithCancelCause(ctx)
+	dctx := &deadlineContext{
+		Context:  cctx,
+		deadline: t,
+	}
+	dur := f.Until(t)
+	if dur <= 0 {
+		dctx.timedOut.CompareAndSwap(false, true)
+		cancelCause(cause)
+		return dctx, func() {
+			cancelCause(context.Canceled)
+		}
+	}
+	stop := f.AfterFunc(dur, func() {
+		if cctx.Err() == nil {
+			dctx.timedOut.CompareAndSwap(false, true)
```

Comment:
I think we can switch this to `.Store(true)`.
The fake clock is meant for use in tests where performance isn't a real concern, so we don't need to be careful about doing a few extra writes to memory.

(I think we should make this substitution generally -- if we needed to know the previous value, etc. then `CompareAndSwap` would make sense, but in the cases here and below, I think it makes sense to do something simpler)